### PR TITLE
Change design for dispatch

### PR DESCRIPTION
**Pull request by dahong67**
Currently we define two methods for `gcp`:
1. https://github.com/dahong67/GCPDecompositions.jl/blob/6c3c18de9da0ec166f96d766404425bcbb5c89a2/src/gcp-opt.jl#L18-L25
2. https://github.com/dahong67/GCPDecompositions.jl/blob/6c3c18de9da0ec166f96d766404425bcbb5c89a2/ext/LossFunctionsExt.jl#L8-L15

These call the following `_gcp` method: https://github.com/dahong67/GCPDecompositions.jl/blob/6c3c18de9da0ec166f96d766404425bcbb5c89a2/src/gcp-opt.jl#L55

**Idea:** unify the two `gcp` methods and pass the inputs directly to `_gcp` so we just have `gcp(X, r, loss = LeastSquaresLoss()) = _gcp(X, r, loss)`. Handle LossFunctions.jl losses by adding methods for `value`, `deriv` and `domain`.

With this design:
+ `gcp` provides the user-friendly interface and handles defaults
+ `_gcp` does not handle defaults - it only implements the algorithms

Separating concerns this way has some potential benefits:
+ reduces duplication in `LossFunctionsExt` - don't need to duplicate code for defaults, etc.
+ prepares for adding specialized `_gcp` methods, e.g., ALS for least-squares (ALS doesn't apply to other loss functions), etc.

### CODE REVIEW

**Comment by dahong67** [comment]
One more TODO remaining:
https://github.com/dahong67/GCPDecompositions.jl/blob/4b5431a4248ef9b50dbe20b9fafb0f16350246cc/src/gcp-opt.jl#L48-L55

Will require reworking how we do the testing since we currently use that signature. For example:
https://github.com/dahong67/GCPDecompositions.jl/blob/4b5431a4248ef9b50dbe20b9fafb0f16350246cc/test/items/gcp-opt.jl#L51-L58

Since this would affect ongoing work in #20, can address this in a separate PR after #20 is merged. Should probably rethink how we do this testing anyway.
